urldac7fd8c97_vitalymaryin_unsafe_utils_tgz-pJ8-org_vit_unsafeutils_TcpTestJ8/src/org/vit/unsafeutils/tcp/TcpStreamConsumer.java

```diff
@@ -0,0 +1,227 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+public class TcpStreamConsumer extends TcpGenericManager implements Service, Runnable {
+    private static Logger logger = Logger.getLogger(TcpStreamConsumer.class);
+    private static boolean debugEnabled = logger.isDebugEnabled();
+    private static final Lock lock = new ReentrantLock();
+    private int selectionOpts = SelectionKey.OP_READ;
+    private boolean connected = false;
+
+    @Override
+    public void stop()  {
+        started = false;
+        if (socket != null && socket.isOpen())
+            try {
+                socket.close();
+            } catch (IOException e) {
+                logger.error(e.getMessage(), e);
+            }
+    }
+
+    @Override
+    public void start(){
+        makeConnection();
+    }
+
+    @Override
+    public boolean shouldStart() {
+        return true;
+    }
+
+    private void makeConnection(){
+        try {
+            hostAddress = InetAddress.getByName(host);
+            selector = SelectorProvider.provider().openSelector();
+            while (socket == null
+                    || ((!socket.isOpen() || !socket.isConnected()
+                    || !socket.socket().isConnected())
+                    && !socket.isConnectionPending())
+                    ){
+                try {
+                    if(socket == null)
+                        socket = openSocket();
+                    else if (!socket.isOpen() || !socket.isConnected()){
+                        socket.close();
+                        socket = openSocket();
+                    }
+                    logger.info("connecting to:" + hostAddress.getHostName() + ":" + port);
+                    boolean success = socket.connect(new InetSocketAddress(hostAddress, port));
+                    if (!success){
+                        while (!success){
+                            success = socket.finishConnect();
+                        }
+                    }
+                    logger.info("connected to:" + hostAddress.getHostName() + ":" + port);
+                    socket.register(selector, selectionOpts);
+                    pendingData.put(socket, new ConcurrentLinkedQueue<byte[]>());
+                    connected = true;
+                }catch (ConnectException e) {
+                    logger.error(e.getMessage(), e);
+                } catch (ClosedChannelException e) {
+                    logger.error(e.getMessage(), e);
+                } catch (ConnectionPendingException e) {
+                    logger.error(e.getMessage(), e);
+                    break;
+                } catch (IOException e) {
+                    logger.error(e.getMessage(), e);
+                }
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)